### PR TITLE
Implement new experimental `fetchpriority` module

### LINKED ISSUE #477: Fetchpriority - add a `fetchpriority=high` attribute to the non-lazy image(s)
## Overview
- Proposed module name: Fetchpriority 
- Proposed module slug: fetchpriority
- Proposed module owner GitHub username(s): pbearne, adamsilverstein

## About the module
Adding the [fetchpriority attribute](https://web.dev/priority-hints/#the-fetchpriority-attribute) `high` attribute to the LCP (or first large) [image](https://web.dev/priority-hints/#increase-the-priority-of-the-lcp-image) on the page should improve the page's LCP, because it tells the browser to prioritize loading that image.

We can build this as a module (possibly experimental), then measure its impact on performance. Note: as of this writing, `fetchpriority` is only supported in [chrome and edge](https://caniuse.com/?search=fetchpriority).
### Scope

WordPress already determines which image is most likely to be the "LCP" image in [`wp_get_loading_attr_default`](https://github.com/WordPress/WordPress/blob/b90c2adb7f65b610c84db80fa35b66cc036e1a66/wp-includes/media.php#L5495-L5516), excluding the `loading=lazy` attribute. We can leverage this to add `fetchpriority=high` to those same images, which should cause the browser to prioritize their load.
### Rationale

`fetchpriority` is a new approach to priority hinting that improves on features like `preload`. Adding this module will let the performance team test this feature as a potential improvement for core.
### Other

Read this document for the details: https://web.dev/priority-hints/#the-fetchpriority-attribute

### PR DESCRIPTION
**Pull request by pbearne**
## Summary

Fixes #477

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

### CODE REVIEW

File: modules/images/fetchpriority/load.php
Comment (adamsilverstein) [inline modules/images/fetchpriority/load.php]:
```suggestion
 * Description: Adds `fetchpriority=high` parameter to the non lazy loaded images.
```

Comment (adamsilverstein) [reply modules/images/fetchpriority/load.php]: is it images or image? typically only one image would be non lazy loaded, right?

File: modules/images/fetchpriority/load.php
Comment (adamsilverstein) [inline modules/images/fetchpriority/load.php]:
```suggestion
 * Filter image tags in content to add fetchpriority to the image tag if lazy loaded is not set.
```

File: modules/images/fetchpriority/load.php
Comment (adamsilverstein) [inline modules/images/fetchpriority/load.php]: interesting, any reference link for this issue?

File: modules/images/fetchpriority/load.php
Comment (adamsilverstein) [inline modules/images/fetchpriority/load.php:32]: I **think** we should only add this at most once (which maybe the case already), so once you add this you could `remove_filter( 'wp_content_img_tag', 'fetchpriority_img_tag_add', 30, 2 );`

File: modules/images/fetchpriority/load.php
Comment (adamsilverstein) [inline modules/images/fetchpriority/load.php]: How did you choose 30?

File: modules/images/fetchpriority/load.php
Comment (mukeshpanchal27) [inline modules/images/fetchpriority/load.php]: Should use `n.e.x.t`.

```suggestion
 * @since n.e.x.t
 * @package performance-lab
```

File: modules/images/fetchpriority/load.php
Comment (pbearne) [reply modules/images/fetchpriority/load.php:32]: good call

File: modules/images/fetchpriority/load.php
Comment (pbearne) [reply modules/images/fetchpriority/load.php]: wanted it to be late 
no good reason 
let me try default 10

File: tests/modules/images/fetchpriority/Fetchpriority-Tests.php
Comment (adamsilverstein) [inline tests/modules/images/fetchpriority/Fetchpriority-Tests.php]: Add an additional test here where the `img` tag contains `loading="lazy"` and the context is `the_content`, verifying the fetchpriority attribute is not added.

Comment (pbearne) [reply tests/modules/images/fetchpriority/Fetchpriority-Tests.php]: added

File: modules/images/fetchpriority/load.php
Comment (pbearne) [reply modules/images/fetchpriority/load.php]: changed

File: modules/images/fetchpriority/load.php
Comment (pbearne) [reply modules/images/fetchpriority/load.php:32]: added

File: modules/images/fetchpriority/load.php
Comment (pbearne) [reply modules/images/fetchpriority/load.php]: no removed comment

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php]: We should make sure to have a space at the beginning of this string, to make sure it doesn't match something like `fooloading="lazy"`. Also, while that's unlikely to already be present, we must avoid adding the `fetchpriority` attribute if the image tag already has one on it, so let's include that here as well.
```suggestion
	if ( strpos( $filtered_image, ' loading="lazy"' ) === false && strpos( $filtered_image, ' fetchpriority=' ) === false ) {
```

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php]:
```suggestion
 * @param string $filtered_image The image tag to filter.
 * @param string $context        The context of the image.
 * @return string The filtered image tag.
```

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php]: Nit-pick, but better to align with core naming:
```suggestion
function fetchpriority_img_tag_add_attr( $filtered_image, $context ) {
```

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php]: See above. Also:
* `remove_filter()` only supports three attributes.
* We need to remove the other filter for featured image which I suggested here too so that it's really always only on the first image it encounters.

```suggestion
		remove_filter( 'wp_content_img_tag', 'fetchpriority_img_tag_add_attr', 10 );
		remove_filter( 'post_thumbnail_html', 'fetchpriority_filter_post_thumbnail_html' );
```

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php]:
```suggestion
add_filter( 'wp_content_img_tag', 'fetchpriority_img_tag_add_attr', 10, 2 );
```

File: tests/modules/images/fetchpriority/Fetchpriority-Tests.php
Comment (felixarntz) [inline tests/modules/images/fetchpriority/Fetchpriority-Tests.php]: See above regarding the function name.

Note also that the test name here is misleading, as this function does not test `wp_filter_content_tags()` at all. It only tests the new function added here, not how it affects `wp_filter_content_tags()`. It also doesn't test that it only adds the attribute to the first image only, so that also needs to be changed in the test name.

Also see my comment above on that we need to add a space before the attribute name when checking for it.

```suggestion
	public function test_fetchpriority_img_tag_add_attr_based_on_context_and_loading_lazy() {
		$attachment_id = $this->factory->attachment->create_upload_object( TESTS_PLUGIN_DIR . '/tests/testdata/modules/images/leafs.jpg' );
		$img           = get_image_tag( $attachment_id, '', '', '', 'large' );

		$this->assertStringContainsString( ' fetchpriority="high"', fetchpriority_img_tag_add_attr( $img, 'the_content' ) );
		$this->assertStringNotContainsString( ' fetchpriority="high"', fetchpriority_img_tag_add_attr( $img, 'not_content' ) );

		$img = str_replace( '<img ', '<img loading="lazy" ', $img );
		$this->assertStringNotContainsString( ' fetchpriority="high"', fetchpriority_img_tag_add_attr( $img, 'the_content' ) );
	}
```

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php]:
```suggestion
 * Filters an image tag in content to add the fetchpriority attribute if it is not lazy-loaded.
```

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php]: One other piece you need to add here is support for the featured image output. You can hook into the `post_thumbnail_html` filter a new callback that just wraps the other callback above. Then you can use the `the_post_thumbnail` context which is already used in WordPress core.

Please add a test for this function as well.
```suggestion
add_filter( 'wp_content_img_tag', 'fetchpriority_img_tag_add', 10, 2 );

/**
 * Filters the post thumbnail HTML to conditionally add the fetchpriority attribute.
 *
 * @since n.e.x.t
 *
 * @param string $html The post thumbnail HTML to filter.
 * @return string The filtered post thumbnail HTML.
 */
function fetchpriority_filter_post_thumbnail_html( $html ) {
	return fetchpriority_img_tag_add_attr( $html, 'the_post_thumbnail' );
}
add_filter( 'post_thumbnail_html', 'fetchpriority_filter_post_thumbnail_html' );
```

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php]: Based on my other comment and code suggestion around the featured image:
```suggestion
	if ( 'the_content' !== $context && 'the_post_thumbnail' !== $context ) {
```

File: tests/modules/images/fetchpriority/Fetchpriority-Tests.php
Comment (felixarntz) [inline tests/modules/images/fetchpriority/Fetchpriority-Tests.php]: We should add another test here that actually calls `wp_filter_content_tags()` to test the integration, since your test above only tests direct calls to the newly added function. For the new test, let's pass some demo content with at least two images in it. The lazy-loading shouldn't matter here. We just need to ensure that the first image has the `fetchpriority` attribute while the second one shouldn't, to test that the removal of the filter works.

Also see my comment above around adding one test for the featured image integration.

**Review by felixarntz — CHANGES_REQUESTED** [review]
@pbearne This is a solid start. I left a few mostly small comments below, but there are two bigger things among them which are critical but missing here:
1. We need to add support for the featured image output, as it is covered by the conditional lazy-loading logic as well.
2. We should add at least one more test to cover the actual `wp_filter_content_tags()` integration of the code here.

File: modules/images/fetchpriority/load.php
Comment (mukeshpanchal27) [inline modules/images/fetchpriority/load.php:32]: If my post contains text and at the end of the article I use an image, then do we also need to add `fetchpriority="high"` attribute? The image is not in view port.

File: modules/images/fetchpriority/load.php
Comment (mukeshpanchal27) [inline modules/images/fetchpriority/load.php:20]: Do we need to add a filter to disable this functionality? Like with Lazy Load, we have `add_filter( 'wp_lazy_loading_enabled', '__return_false' );` filter in the core.

**Review by mukeshpanchal27 — COMMENTED** [review]
@pbearne great work. It look good to me. Left some questions.

**Comment by mukeshpanchal27** [comment]
https://github.com/WordPress/performance/pull/528/commits/5480988b0e66b5d07cd93944f32fe95c471c8117 Use lowercase file name for test class file commit reverted in https://github.com/WordPress/performance/pull/528/commits/e6826eb1f164ede4582d793d709d2e925540e8d4 that needs to be correct.

Do we need `/tests/modules/images/fetchpriority` in CODEOWNERS?

File: modules/images/fetchpriority/load.php
Comment (pbearne) [reply modules/images/fetchpriority/load.php:32]: I am letting the loading="lazy" code work if to add the as don't what to have both this is an else, and I suspect that when add it to code, it will be mixed into the code that adds loading="lazy"

File: modules/images/fetchpriority/load.php
Comment (pbearne) [reply modules/images/fetchpriority/load.php:20]: not here, but in core yes.
we can just turn it off in the plugin

**Comment by mukeshpanchal27** [comment]
@bethanylang The PR needs a `Type` label. PR failing now because of it.

**Comment by mxbclang** [comment]
@mukeshpanchal27 Ah, that's right! Fixed.  :)

File: modules/images/fetchpriority/load.php
Comment (adamsilverstein) [inline modules/images/fetchpriority/load.php:54]: Nice!

File: tests/modules/images/fetchpriority/fetchpriority-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/fetchpriority/fetchpriority-tests.php]: Remove space

File: tests/modules/images/fetchpriority/fetchpriority-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/fetchpriority/fetchpriority-tests.php]: Test is for `fetchpriority_img_tag_add_attr` function isn't it? And should use lowercase for `@covers` annotation.

```suggestion
 * @covers ::fetchpriority_img_tag_add_attr
```

File: tests/modules/images/fetchpriority/fetchpriority-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/fetchpriority/fetchpriority-tests.php:41]:
```suggestion
		$attachment_id = $this->factory->attachment->create_upload_object(
			TESTS_PLUGIN_DIR . '/tests/testdata/modules/images/leafs.jpg'
		);

		$img = get_image_tag( $attachment_id, '', '', '', 'large' );
```

File: tests/modules/images/fetchpriority/fetchpriority-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/fetchpriority/fetchpriority-tests.php]: Should remove this as we get `$attachment_id` from other code. Check other comment.

**Review by mukeshpanchal27 — COMMENTED** [review]
@pbearne Left some review for the unit test.

**Comment by mxbclang** [comment]
@pbearne Just a reminder on changes requested here, thank you!

File: tests/modules/images/fetchpriority/fetchpriority-tests.php
Comment (pbearne) [reply tests/modules/images/fetchpriority/fetchpriority-tests.php]: code removed

**Comment by pbearne** [comment]
I need help getting the tests to work.
I expect the first image to not have to lazyload set, but I am getting this so tests are failing

**Comment by gilbertococchi** [comment]
Thanks Paul for all the efforts into building this module.

Were you able to identify in what use case the attribute loading=lazy is added on the first image?

I also was able to reproduce that and got some feedback from WP developers that this happens but I wasn't able to find the root cause in WP Core just yet.

I can see how this might undermine the potential impact of adding the fetchpriority if there are critical cases where it would not be added.

**Comment by pbearne** [comment]
Look at the core code It should miss the first image loaded on a page it uses this function track the count wp_increase_content_media_count()
You can see an example in wp_get_attachment_image()

It may be working on a site but not in the tests

**Comment by gilbertococchi** [comment]
Thanks for your answer @pbearne, I can totally get why the test would fail in the tests, @felixarntz FYI.

I will investigate on other threads why WP Sites apply so often the `loading=lazy` value on first images, that was one of my concerns to use the same logic to decide when to apply the `fetchpriority` attribute as well.

**Comment by adamsilverstein** [comment]
@pbearne I'm going to see if I can tell why the tests are failing here. Were you able to resolve all the other previous feedback (I see some unresolved comments in the thread)?

File: modules/images/fetchpriority/load.php
Comment (adamsilverstein) [reply modules/images/fetchpriority/load.php:32]: > If my post contains text and at the end of the article I use an image, then do we also need to add fetchpriority="high" attribute? The image is not in view port.

The short answer is no. However, we are expecting you will benefit from marking your first image - based on the expectation that on balance, it will be your LCP element.

File: tests/modules/images/fetchpriority/fetchpriority-tests.php
Comment (adamsilverstein) [reply tests/modules/images/fetchpriority/fetchpriority-tests.php]: removed in incoming code

File: tests/modules/images/fetchpriority/Fetchpriority-Tests.php
Comment (adamsilverstein) [reply tests/modules/images/fetchpriority/Fetchpriority-Tests.php]: this is already updated in the code base

File: modules/images/fetchpriority/load.php
Comment (adamsilverstein) [reply modules/images/fetchpriority/load.php]: this is in place in the codebase

File: tests/modules/images/fetchpriority/Fetchpriority-Tests.php
Comment (adamsilverstein) [reply tests/modules/images/fetchpriority/Fetchpriority-Tests.php]: this is in place as `test_fetchpriority_img_tag_add_in_wp_filter_content_tags`

**Comment by pbearne** [comment]
@adamsilverstein I have lost track of where we all are. If you can sort the tests/code, I will revisit and resolve the tickets 
I suspect that the issue is with the lazyloading code, not this code as we are piggybacking on that

**Comment by adamsilverstein** [comment]
@pbearne I tracked down the issue(s) with the tests and got them all passing:
- the filter in core only does the counting & limiting lazy from first image for images that are "in the loop" and where `is_main_query` is true (a restriction that is perhaps too limiting?) so I had to mock that condition
- the duplicate image id caused an issue because `wp_filter_content_tags` has an optimization that prevents it from repeating logic on identical images.Using two separate image Ids fixed the issue.
I also added some cleanup logic for the attachments.

Please review and we can try to land this soon.

**Comment by adamsilverstein** [comment]
I verified this works as expected in a classic theme like twentytwelve.

when testing with twentytwentythree however, I noticed that the first image is lazy loaded, so something isn't quite working with the logic in core for block themes. I'm going to dig into this a bit further to see why. cc: @felixarntz

**Review by adamsilverstein — APPROVED** [review]
LGTM

**Comment by adamsilverstein** [comment]
> when testing with twentytwentythree however, I noticed that the first image is lazy loaded, so something isn't quite working with the logic in core for block themes. I'm going to dig into this a bit further to see why. 

As we had discussed earlier, this is caused by the block theme "double" pass at the content (which can help performance). To fix this we need to avoid "counting" the first pass by block themes. I noticed in this case the `$context` is false, so arrived at this simple solution which fixes the issue in my testing: https://github.com/WordPress/wordpress-develop/pull/3538

I will open up a Trac ticket to discuss further.

**Comment by adamsilverstein** [comment]
@pbearne this is ready for review.

**Comment by adamsilverstein** [comment]
@felixarntz this is ready for review

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php:31]: Since we rely on the presence of `loading="lazy"` here, we first of all have to make sure the site has lazy-loading enabled in general (via `wp_lazy_loading_enabled( 'img', $context )`). Otherwise, this will cause a bug on sites that disabled it, effectively giving every image `fetchpriority="high"` which would be detrimental. Let's add a test to cover that scenario as well.

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php:33]: I think for safety reasons we need to rather include the other filter removal call here as well. Essentially, whenever we add `fetchpriority="high"` to any image, we need to remove all other filters that could possibly do it, so that it's always just one image.

Right now, the way the code is written it would be possible in principle for 2 images to have `fetchpriority="high"` when a featured image without `loading` attribute follows a content image without `loading` attribute. Unlikely but possible, and we need to be stricter here than with lazy-loading as preferably really only the LCP image should have `fetchpriority="high"`.
```suggestion
		remove_filter( 'wp_content_img_tag', 'fetchpriority_img_tag_add_attr' );
		remove_filter( 'post_thumbnail_html', 'fetchpriority_filter_post_thumbnail_html' );
```

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php]: See above, we should then remove the call here in favor of the one above.
```suggestion
```

**Review by felixarntz — CHANGES_REQUESTED** [review]
@pbearne @adamsilverstein Overall this looks solid, though there are a few things we need to fix as they can introduce bugs.

In light of the currently ongoing work to fix image lazy-loading in block themes (see https://core.trac.wordpress.org/ticket/56930), we should make sure to test prior to merge that this works as expected in a block theme.

Looking at the code here, things look like this may not be affected by the double execution bug from core (due to the more explicit context check present here), but we need to verify that. Adding `fetchpriority="high"` to too many images may be even worse than lazy-loading too many images.

File: tests/modules/images/fetchpriority/fetchpriority-tests.php
Comment (felixarntz) [inline tests/modules/images/fetchpriority/fetchpriority-tests.php]: What is the purpose of this? To check how often `fetchpriority` is included in the content? In that case we should rather use `substr_count`. This logic looks confusing :)

Comment (adamsilverstein) [reply tests/modules/images/fetchpriority/fetchpriority-tests.php]: i was confused at first as well, this is just counting that fetchpriority is included exactly once, agree changing this to a substr_count makes more sense 👍🏼

File: modules/images/fetchpriority/load.php
Comment (adamsilverstein) [reply modules/images/fetchpriority/load.php:31]: 👍🏼 added a check above @ L26

File: tests/modules/images/fetchpriority/fetchpriority-tests.php
Comment (adamsilverstein) [reply tests/modules/images/fetchpriority/fetchpriority-tests.php]: Adjusted

**Comment by adamsilverstein** [comment]
@felixarntz Addressed feedback and ready for another review

**Comment by adamsilverstein** [comment]
> we should make sure to test prior to merge that this works as expected in a block theme.

I will double check this.

**Comment by felixarntz** [comment]
Milestoning this for the 1.8.0 release coming in December, as it's shaping up for merge by then. I'll give it another review later this week.

File: modules/images/fetchpriority/load.php
Comment (felixarntz) [inline modules/images/fetchpriority/load.php]: This description is a bit too technical and too close to the detailed implementation. Also, we can't use Markdown since this will be displayed as is in HTML. I suggest something like:
```suggestion
 * Description: Adds a fetchpriority hint for the primary content image on the page to load faster.
```

**Review by felixarntz — APPROVED** [review]
@pbearne @adamsilverstein The implementation here LGTM, with the following caveats / known issues:
* For block themes, it currently doesn't work on featured images since the featured image block is broken in regards to `loading="lazy"`.
    * This was recently fixed in Gutenberg (see https://github.com/WordPress/gutenberg/pull/45534).
* For block themes, it already works for regular content images, however the first content image will still have `loading="lazy"` on it due to the double content parsing bug in core.
    * A WordPress core fix is being worked on (see https://github.com/WordPress/wordpress-develop/pull/3560).

Both of these problems are not critical, since at least this PR doesn't make it worse than it is; and both of these problems need to be fixed separately, so this is good to merge IMO.

Will update the module description though, as it's a bit too technical and doesn't describe the performance impact.

**Comment by felixarntz** [comment]
FYI the PR https://github.com/whatwg/html/pull/8470 is where adding `fetchpriority` to the HTML spec is being worked on. This will be critical before we can eventually merge it into WordPress core. But for the plugin shipping this already should be fine.

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks @pbearne @adamsilverstein LGTM

**Pull request merged**
